src/sage/categories/finite_fields.py

```diff
@@ -243,5 +245,283 @@ def _element_of_factored_order(self, F):
+
+            - A square root of the element; raises an error
+              if the element is not a square
+
+            EXAMPLES::
+                sage: k.<a> = GF((5, 10))
+                sage: k(2).is_square()
+                True
+                sage: k(2)._cipolla()^2 == k(2)
+                True
+                sage: k.quadratic_non_residue()._cipolla()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+            """
+            parent = self.parent()
+            q = parent.cardinality()
+            if not self.is_square():
+                raise ValueError("Element is not a square")
+            t = parent.random_element()
+            root = t**2 - 4 * self
+            while root.is_square():
+                t = parent.random_element()
+                root = t**2 - 4 * self
+            from sage.rings.polynomial.polynomial_ring import polygen
+            X = polygen(parent)
+            f = X**2 - t*X + self
+            b = pow(X, (q+1)//2, f)
+            return b
+
+        def sqrt(self, all=False, algorithm='Tonelli'):
+            """
+            Returns the square root of the element if it exists
+
+            INPUT:
+
+            - ``all`` -- boolean (default: ``False``); whether to return a list of
+              all square roots or just a square root
+
+            - ``algorithm`` -- string (default: 'Tonelli'); The algorithm to use
+              among ``'Tonelli'``, ``'Cipolla'``. Most finite fields will have
+              Tonelli to run fast, but in the case where the order minus 1 is
+              highly divisible by 2 Cipolla's method could be faster.
+
+            OUTPUT:
+
+            - if ``all=False``, a square root; raises an error if the element is not
+              a square
+
+            - if ``all=True``, a 2-tuple of square roots; raises an error if the
+             element is not a square
+
+            EXAMPLES::
+                sage: S.<x> = GF(5)[]
+                sage: f = S.irreducible_element(20)
+                sage: k.<y> = S.quotient_ring(f)
+                sage: k in Fields()
+                True
+                sage: k(2).is_square()
+                True
+                sage: k(2).sqrt()^2 == k(2)
+                True
+                sage: k(4).sqrt(all=True)
+                (2, 3)
+                sage: k.quadratic_non_residue().sqrt()
+                Traceback (most recent call last):
+                ...
+                ValueError: Element is not a square
+
+            ALGORITHM:
+
+            The following algorithm comes from chapter 8 of [BS1996]_.
+
+            If `q = p^n` is the order of the finite field then if
+            `p = 2` or divisible by 2 then we can compute the
+            square root by `a^(q/2)` if q is 3 modulo 4 then we
+            can compute the square root by `a^((q+1)/4)` otherwise
+            we use tonelli's method for all other cases in general.
```

Comment:
```suggestion
            The algorithms used come from chapter 6 of [BS1996]_.
            Let `q = p^n` be the order of the finite field, let `a` be the finite field element
            that we wish to find the square root of.
            
            - If `p = 2` then `a` is always a square, and the square root of `\sqrt{a} = a^{(q / 2)}`.
            - If `q \equiv 3 \pmod{4}` then if `a` is a square `\sqrt{a} = {a^((q+1) / 4)}`
            - For all other cases we use the algorithm given by the ``algorithm`` parameter.
```

Wording, I need to check that this formatting is okay in the generated HTML.